File: go/unionfind.go
package algorithm

import "fmt"

type UnionFindType struct {
	vToNodeMap map[int]*unionFindNodeType
}

func UnionFind() *UnionFindType {
	return &UnionFindType{
		vToNodeMap: make(map[int]*unionFindNodeType),
	}
}

// Add one node
func (uf *UnionFindType) Add(v int) error {
	if uf.vToNodeMap[v] != nil {
		return fmt.Errorf("value %d already existed in UnionFind", v)
	}
	uf.vToNodeMap[v] = unionFindNode(v)
	uf.vToNodeMap[v].root = uf.vToNodeMap[v]
	return nil
}

func (uf *UnionFindType) Find(v int) (int, error) {
	if uf.vToNodeMap[v] == nil {
		return -1, fmt.Errorf("value %d not found in UnionFind", v)
	}
	return uf.find(uf.vToNodeMap[v]).val, nil
}

func (uf *UnionFindType) Union(a, b int) {
	if uf.vToNodeMap[a] == nil {
		uf.Add(a)
	}
	if uf.vToNodeMap[b] == nil {
		uf.Add(b)
	}
	uf.union(uf.vToNodeMap[a], uf.vToNodeMap[b])
}

// union by rank
func (uf *UnionFindType) union(a, b *unionFindNodeType) {
	aRoot := uf.find(a)
	bRoot := uf.find(b)
	if aRoot != bRoot {
		var small, large *unionFindNodeType
		if aRoot.rank >= bRoot.rank {
			small = bRoot
			large = aRoot
		} else {
			small = aRoot
			large = bRoot
		}
		small.root = large
		if small.rank == large.rank {
			large.rank++
		}
	}
}

func (uf *UnionFindType) find(node *unionFindNodeType) *unionFindNodeType {
	if node == node.root {
		return node
	}
	node.root = uf.find(node.root)
	return node.root
}

type unionFindNodeType struct {
	val  int
	root *unionFindNodeType
	rank int
}

func unionFindNode(val int) *unionFindNodeType {
	return &unionFindNodeType{
		val:  val,
		root: nil,
		rank: 0,
	}
}

File: cpp/src/union_find.cpp
#include "union_find.hpp"

void UnionFind::add(int v)
{
    if (m_nodeMap.find(v) != m_nodeMap.end()) {
        printf("Node %d already existed", v);
        return;
    }
    m_nodeMap[v] = std::make_shared<Node>(v);
    m_nodeMap[v]->root = m_nodeMap[v];
}

int UnionFind::find(int v)
{
    if (m_nodeMap.find(v) == m_nodeMap.end()) {
        printf("Node %d does not found", v);
        return -1;
    }
    return _find(m_nodeMap[v])->value;
}

void UnionFind::group(int a, int b)
{
    if (m_nodeMap.find(a) == m_nodeMap.end()) {
        add(a);
    }
    if (m_nodeMap.find(b) == m_nodeMap.end()) {
        add(b);
    }
    return _group(m_nodeMap[a], m_nodeMap[b]);
}

std::shared_ptr<UnionFind::Node> UnionFind::_find(std::shared_ptr<UnionFind::Node> node)
{
    if (node == node->root) {
        return node;
    }
    node->root =  _find(node->root);
    return node->root;
}

void UnionFind::_group(std::shared_ptr<UnionFind::Node> a, std::shared_ptr<UnionFind::Node> b)
{
    auto small = _find(a);
    auto large = _find(b);

    if (small->rank > large->rank) {
        std::swap(small, large);
    }

    small->root = large;

    if (small->rank == large->rank) {
        large->rank++;
    }
}